### ISSUE TITLE
Properties or public variables

### ISSUE THREAD

**Issue by andreaswitzen** [andreaswitzen/TopDownSurvival#5]

**Comment by andreaswitzen** [comment]
You can write "good" code by using [serializefield] instead of public to make a field visible in the editor inspector, and then use C# properties to access them from elsewhere in script. Is this extra work always worth it? Are there cases when just using public variables are okay?

**Comment by canisin** [comment]
As far as I know the reason that properties are preferable to fields is a) separate access control for read and write and b) future proofing in case there is a change to how the value is stored. C# makes the whole thing very much painless by providing lots of syntactic shortcuts for working with properties like auto properties, etc.

**Comment by andreaswitzen** [comment]
I am aware of the benefits and also remember that you thought properties is the way to go when we discussed this a while back. After that I stuck with properties for a while but found them cumbersome in some cases. For instance, to make something editable in the inspector you also need a explicit private field.

Question is, are there cases when a public field is as good or better? E.g. simple database objects that only store data?

**Comment by canisin** [comment]
After a quick google search, it seems that unity really does not want to work with properties, so i think we can just go ahead and use public fields for what we want to expose to the editor.

It might make sense in that anything that is to be exposed by the editor has to have a known memory location.

Moreover, now that i think about it, all the solutions that i saw about exposing properties to unity editor via attributes, would likely fail if the property was anything other than an automatic property with a backing field.

Let me try to clarify with an example. Say your class has two properites a Foo and a Bar and there is a constant relationship between these two where Bar is always twice the value of Foo. Imagine that providing both is meaningful because of real world connotations, for example, mass and weight.

You would likely code these properties as

public float Foo { get; set; }
public float Bar { get { return Foo * 2.0; } set { Foo = value / 2.0; } }

I don't think we would be able to expose Bar to the unity editor with any of the solutions that i saw on my google search.

**Comment by canisin** [comment]
> Question is, are there cases when a public field is as good or better?
I would say a public field is better when you want to expose a value to the unity editor! :D

**Comment by andreaswitzen** [comment]
Ok! Sounds good to me as you can also have fields be private but still visible in the inspector with the [serializefield] tag. As you can tell from my comment above my main problem with properties is the same as the one you bring up.

**Comment by canisin** [comment]
I cannot immediately think of a reason why a field should both be private and exposed. We should not have any need of the attribute if we stick to making editor fields public.

**Comment by andreaswitzen** [comment]
It depends on how much you like to rely on the editor. Over time I have come to prefer relying as much as possible on setting things up in the editor over writing script to initialize things.

As a (somewhat) concrete example you can set up a player's or enemy's starting inventory through the editor but you probably don't want to expose the inventory list/array to other classes, as this increases the risk for bugs (safer to only allow adding items through the inventory's Add() and Remove() methods).

**Comment by andreaswitzen** [comment]
Another benefit of [serializefield] is that only the fields that you actually want to edit in the inspector are visible in the inspector. If you go with making all fields public you will have a lot of fields visible in the inspector that aren't meant to be edited there (more clutter, more potential for making mistakes, project gets harder to understand over time).

**Comment by canisin** [comment]
> As a (somewhat) concrete example you can set up a player's or enemy's starting inventory through the editor but you probably don't want to expose the inventory list/array to other classes, as this increases the risk for bugs (safer to only allow adding items through the inventory's Add() and Remove() methods during runtime).

hmm.. that kind of does not compute. if there is any reason that adding or removing an item from the player's inventory should go thru the Add/Remove functions, then that same reason should apply for the editor as well. And conversely, if it is ok for the editor to directly insert into the inventory list then that should be ok for the other classes as well.

**Comment by canisin** [comment]
> making all fields public

no, let's not make _all_ fields public. let's just make those fields that we won't to expose to the editor public and accept that they are part of the class's interface.
